### PR TITLE
Properly setup seed in RNG

### LINKED ISSUE #27166: RandomPCG::seed does not properly seed PCG32
**Godot version:** 3.1 stable, still in df7d370 (current master)

**OS/device including version:** N/A

**Issue description:** RandomPCG::seed is different from PCG32's original seeding function. As such, the statistical performance of the original PCG32 cannot be assumed for the current implementation.

Godot's random_pcg.h:

```cpp
_FORCE_INLINE_ void seed(uint64_t p_seed) {
	current_seed = p_seed;
	pcg.state = p_seed;
	pcg32_random_r(&pcg); // Force changing internal state to avoid initial 0
}
```

Actual srandom function from PCG's C minimal implementation: https://github.com/imneme/pcg-c-basic/blob/master/pcg_basic.c

```c
// Copyright 2014 Melissa O'Neill <oneill@pcg-random.org>, Licensed under the Apache License, Version 2.0
void pcg32_srandom_r(pcg32_random_t* rng, uint64_t initstate, uint64_t initseq)
{
    rng->state = 0U;
    rng->inc = (initseq << 1u) | 1u;
    pcg32_random_r(rng);
    rng->state += initstate;
    pcg32_random_r(rng);
}
```

**Steps to reproduce:**  N/A

**Minimal reproduction project:** N/A

### PR DESCRIPTION
**Pull request by Chaosus**
Looks like I've been hurried when creating #25680 and didnt used a native solution which is usage of pcg32_srandom_r specially created for this purpose in http://www.pcg-random.org/downloads/pcg-c-basic-0.9.zip

Fix #27166 + changed few useless/invalid code parts

### CODE REVIEW

**Review by hpvb — DISMISSED** [review]
If we're adding more licensed code from a third party should this go into third-party/README and/or the license metadata file? @akien-mga?

File: core/math/random_pcg.h [hunk 1]
```diff
@@ -38,18 +38,18 @@
...
 	static const uint64_t RANDOM_MAX = 0xFFFFFFFF;
 
-	RandomPCG(uint64_t p_seed = DEFAULT_SEED, uint64_t p_inc = PCG_DEFAULT_INC_64);
+	RandomPCG(uint64_t p_seed = DEFAULT_SEED, uint64_t p_inc = DEFAULT_INC);
 
 	_FORCE_INLINE_ void seed(uint64_t p_seed) {
 		current_seed = p_seed;
-		pcg.state = p_seed;
-		pcg32_random_r(&pcg); // Force changing internal state to avoid initial 0
+		pcg32_srandom_r(&pcg, current_seed, current_inc);
```
Comment (toasteater) [inline core/math/random_pcg.h]: From http://www.pcg-random.org/using-pcg-c-basic.html:

> For this generator, there are 2^63 possible sequences of pseudorandom numbers. Each sequence is entirely distinct and has a period of 2^64. The initseq argument selects which stream you will use. The initstate argument specifies where you are in that 2^64 period.

Maybe `current_inc` should be renamed `current_seq` to better reflect it's usage? Actual `inc` from pcg32_srandom_r is `(initseq << 1u) | 1u` so this might be a bit confusing.

**Comment by Chaosus** [comment]
@hpvb Its code from the same lib, I doubt that even -- GODOT -- start/end needs to be required here, I guess pcg.h/c was manually created, 2 years ago by Ferenc Arn

File: core/math/random_pcg.h (hunk 1 above)
Comment (Chaosus) [reply core/math/random_pcg.h]: Its also required to change `pcg32_random_t `cuz its have `inc` inside. I think its redundant, and its not a target for this PR

File: thirdparty/misc/pcg.cpp [hunk 2]
```diff
@@ -13,3 +13,14 @@ uint32_t pcg32_random_r(pcg32_random_t* rng)
     uint32_t rot = oldstate >> 59u;
     return (xorshifted >> rot) | (xorshifted << ((-rot) & 31));
 }
+
+// -- GODOT start --
```
Comment (akien-mga) [inline thirdparty/misc/pcg.cpp]: I think there's no need to add the `// -- GODOT start --` comments here as it's code from the same upstream. Just add a link to its source so that we understand where these two methods were taken from.

Comment (Chaosus) [reply thirdparty/misc/pcg.cpp]: Yeah, I think the same. Where to add this link ? Like so ? 
image

Comment (Chaosus) [reply thirdparty/misc/pcg.cpp]: Ok, changed it like the picture above

**Comment by Chaosus** [comment]
Fixed it to original look (without modify by clang-format)

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.1.1.

**Comment by MidZik** [comment]
The change is fine but it's going to break backwards compatibility again, seed will produce different sequences than in 3.1.0.

**Comment by akien-mga** [comment]
Indeed, thanks. I'll revert the 3.1 cherry-pick.

**Comment by MidZik** [comment]
Actually I just realized that this change breaks saving/restoring the RNG state again. The seeding function no longer treats the seed as a state, so after calling `rand()` once, passing `get_seed()` into `seed()` is not going to put the RNG into the same state. In fact there is no way to get a valid "seed" out of the state anymore.

This change needs `get_state()` and `set_state()` methods to go along with it, and `get_seed()` would need to either be removed entirely, or changed to return the _initial_ seed, which would only change if you re-seeded the RNG.

**Comment by Chaosus** [comment]
>  so after calling rand() once, passing get_seed() into seed() is not going to put the RNG into the same state

I think this can be simply fixed by removing `current_seed = pcg_state` in `rand()`

**Comment by toasteater** [comment]
> The change is fine but it's going to break backwards compatibility again, seed will produce different sequences than in 3.1.0.

Sure, but should that be considered a compatibility break? It's not like RandomNumberGenerator's implementation is spec-frozen like STL's 9000 compatibility RNGs that differ only in constants. A lot of languages don't guarantee that their standard random implementation never changes across versions, and I can't see why Godot's should. If you know you what you want, you should generate numbers with code that *you* control, no?

However, I do see that reliable random numbers is a common need in gamedev. Maybe we should introduce a frozen PRNG (alongside the current RNG class, under another name) whose behavior will absolutely never change despite how buggy it is? That will give users like @MidZik something to use.

**Comment by MidZik** [comment]
> 
> 
> > so after calling rand() once, passing get_seed() into seed() is not going to put the RNG into the same state
> 
> I think this can be simply fixed by removing `current_seed = pcg.state` in `rand()`

What does that fix? You still can't save/restore the RNG state.

> Sure, but should that be considered a compatibility break?

Yes. The reproducibility of PRNGs is one of their major features, and I don't think their behavior should change in a _revision_ release, which generally aims to keep behavior the same where possible.

**Comment by avencherus** [comment]
@toasteater It is a legitimate concern if your game is of a procedurally generated kind, and the seed number is used to reproduce a level.  If the version changes, and the seed evolves differently, that level cannot be recreated anymore from one version to the next.  You may also be exposing this feature to players to share with one another.  Rather suddenly, if you hadn't seen this kind change, you may a small disaster on your hands.

Factorio is a great example of this kind of usage:  https://wiki.factorio.com/World_generator

Many people share and pass their world seeds around.

Also in the case of game replay systems, which may be for user enjoyment or for debugging and supporting a game, these will no longer work on old data. 

If updates to seed sequences are too frequent, a creator of a game that relies on these uses would find themselves locked to a specific version, only for this kind of change.

The use cases are of a more specific nature, but it isn't necessarily trivial, certainly not one where you might change things without any reflection.  But I wouldn't disagree with you that the RNG of a framework shouldn't be ossified in stone either.

I like the improvements you've been doing, but I can also be counted as one hoping that certain aspects hit a stable point soon, because I know I'm going to be attempting to make use of replays in my games in the not too distant future.  For me though I'm mostly creating a custom version, so it wouldn't be too much of a problem to revert specific commits.

**Comment by Chaosus** [comment]
Looks like a serious problem, I prefer to revert it back

**Comment by toasteater** [comment]
> The reproducibility of PRNGs is one of their major features

The most major feature of PRNGs are their randomness.

> It is a legitimate concern if your game is of a procedurally generated kind, and the seed number is used to reproduce a level.
> ...
> I can also be counted as one hoping that certain aspects hit a stable point soon

I did acknowledge that stability is a legitimate concern, and I agree with you. That's why I'm suggesting that for these uses you generate numbers with code that you control, and that a frozen PRNG can be introduced alongside the default one.

**Comment by Zireael07** [comment]
Reproducibility and seedability of PRNGs is nearly as important as randomness.

**Comment by avencherus** [comment]
> A lot of languages don't guarantee that their standard random implementations never change across versions, and I can't see why Godot's should. If you know you what you want, you should generate numbers with code that you control, no?

@toasteater I see, yeah disregard what I wrote.  X)  I misunderstood you a bit because of these remarks above.

The recommendation would be good, though I'm not sure what the maintainers approach would be here with the desire for backwards compatibility.  Maybe new versions become additional API, at least until major releases can make things obsolete/unsupported?  But yes, older versions would be good to keep around.

I'm not sure exactly the extent of the bugginess you've mentioned, but I am curious.  I'm not so well versed in this, and so far I haven't noticed anything clearly detrimental in my uses in the recent past.

**Comment by toasteater** [comment]
@avencherus I'm sorry if my comment was inappropriate in tone, English isn't my first language, but I'm glad that we came to understand each other now. I agree that this is best left for the maintainers to handle.

As for the extent of bugginess (for this specific issue), here's some background: PCG32, the RNG that Godot is currently using, can be described as an LCG with a complex output function. The output function is impressive, but the LCG part is still susceptible to "bad state" -- that's why it once had the ["initial 0" bug](https://github.com/godotengine/godot/issues/17664).  #25680 masked the problem, but one can still easily find "bad seeds" for it (`0`, for a trivial example). Without proper initialization, the RNG isn't really PCG32, and all the glowing reviews of PCG32's randomness simply don't apply. If users *assume* it's good, then bizarre problems can plague them somewhere downstream.

The same can be said of issues with it's distributions: 4 billion is large until it isn't. randi_range seem benign until you start dealing with millions, then the bias becomes devastating. Of course, the average mobile developer will not care, but it fails exactly the moment it becomes important, and that is the problem.

**Comment by avencherus** [comment]
@toasteater Thanks.  I'll likely be cherry-picking the improvements to the RNG up until some version, so I may not experience it this bad initialization issue.  Are there side effects aside from the initial 0 number issue?  Are the distributions getting clumped or something like that?

**Comment by toasteater** [comment]
> Are there side effects aside from the initial 0 number issue? Are the distributions getting clumped or something like that?

Yes, the distribution may stop being uniform for some seeds<sup>*</sup>. The initial 0 is just the most obvious symptom: PCG32's output function does a good job at spreading the distribution out, but it's still a garbage in, garbage out situation.

If you are using `rand()` directly, and are choosing good, random seeds, then it's more likely that you'll not notice anything outright biased to the naked eye. If you're using [randi_range](https://github.com/godotengine/godot/issues/27094), [randf](https://github.com/godotengine/godot/pull/27193), or randfn (uses randf internally) though, the current implementations introduce their own biases that can be significant or not depending on your use case.

Bottom line: avoid the current implementation for any simulations (e.g. board game AIs), or in use cases where fairness is important (multiplayer, etc.)

- - -

<sup>*</sup> Technically, it is "we can't say with confidence that the distribution is as uniform as PCG32's, which is believed to be uniform".
